tfjs-layers/src/layers/core.ts

```diff
@@ -673,3 +738,344 @@ export class Masking extends Layer {
   }
 }
 serialization.registerClass(Masking);
+
+export interface TransformerLayerArgs extends LayerArgs {
+  numHeads: number;
+  depth: number;
+  pool: boolean;
+  padSize: number;
+}
+
+class TransformerLayer extends Layer {
```

Comment:
Quick comment: TransformerLayer doesn't seem to be in the standard TensorFlow (Python) API. We only add layer types that are already in TensorFlow (Python) to the TF.js layers API.

Reply:
Hi @caisq, thanks for taking time to read my PR ! 

Would it be better if I respect the layer named [MultiHeadAttention](https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention) (which does the same thing) ?  

It would only be minor changes. 

However, I'm still not sure how to access the layers I need to make it build in `tfjs`.

Thanks for the feedback

Reply:
Yes, implementing MultiHeadAttention will be good. Please note that we generally strive for high fidelity in replicating the behavior of Python. The implementation of MultiHeadAttention in Python uses the einsum op, which has been recently added to tfjs-core. But note that because the gradient of the einsum op hasn't been implemented in tfjs-core yet, the first implementation of MultiHeadAttention in tfjs most likely won't be able to support training.

Reply:
Hi @caisq, thanks a lot for the hindsights. 

Could we start with `Dense` only and then create another PR when einsum will be stable for training ? Or is the python fidelity more important ? 

In any case I will start to change the code to match the `MultiHeadAttention` behaviour. 

Note that I'm doing this on my free time so I will comment here when modifications are made. 

Have a great day.

Reply:
Not sure what you mean by "start with `Dense`". `Dense` is already implemented in TF.js. 

I still there there is value in adding `MultiHeadAttention` to TF.js at this time, because there is no specialized code for training in the code for that layer. So when einsum gradient is supported in tfjs-core in the future, training for that layer should just work. Before the gradient is supported, the layer can still support inference.

Reply:
Sorry if I wasn't clear, I meant using classic fully connected layers to create the MultiheadAttention so we can train it and then refactor when einsum gradient will be available. 

Then I understand that you still prefer to do everything with einsum from now on even if we cannot train it. 

So I will change the code to use einsum, do you have any ETA on when the gradient will be available ?

Reply:
Sorry for the delay in response. But my take is to use the einsum implementation. This will involve less total amount of work. I don't have an ETA for the gradient of einsum yet. But einsum grad is not terribly hard to write.

Reply:
I replaced the `K.dot(a,b, bias)` with `add(eisum('eq', a,b))` as required. 

However I don't know if there is an equation to do the dot and the sum at once. 

To be honest I find harder to read einsum than plain methods like `.add` or `.dot` why is this better ? 

The only pain point is the `layerNormalisation`, I don't know how to get it and apply it in the `tfjs-layers/core.ts` as it's another layer...

Then just back and forth to ensure everything is correct and match your guidelines/API requirements. 

Thanks again for your time on this PR. 

Have a great day.